infrastructure/src/main/java/org/corfudb/infrastructure/log/Segment.java

```diff
@@ -0,0 +1,557 @@
+package org.corfudb.infrastructure.log;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.InvalidProtocolBufferException;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.corfudb.common.metrics.micrometer.MicroMeterUtils;
+import org.corfudb.infrastructure.ResourceQuota;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.exceptions.DataCorruptionException;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.corfudb.infrastructure.log.SegmentUtils.getByteBuffer;
+import static org.corfudb.infrastructure.log.SegmentUtils.getLogData;
+import static org.corfudb.infrastructure.log.SegmentUtils.getLogEntry;
+import static org.corfudb.infrastructure.log.SegmentUtils.getMetadata;
+import static org.corfudb.infrastructure.log.SegmentUtils.getSegmentHeader;
+import static org.corfudb.infrastructure.utils.Persistence.syncDirectory;
+
+/**
+ *
+ * A segment is a dense range of sequence numbers that are mapped to log entries. Each segment
+ * is backed by a single file and maintains a mapping from (sequence number) -> (file offset, payload length).
+ * A segment maintains an in-memory hash index to file locations to speed up random reads and lookups.
+ *
+ * The concurrency model supported by a segment is a single writer, multiple readers model where
+ * write operations must be owned by a single thread, while multiple threads can concurrently do
+ * read operations.
+ *
+ * @author Maithem
+ */
+@Slf4j
+public class Segment {
+    public static final int METADATA_SIZE = LogFormat.Metadata.newBuilder()
+            .setLengthChecksum(-1)
+            .setPayloadChecksum(-1)
+            .setLength(-1)
+            .build()
+            .getSerializedSize();
+
+    public static final int VERSION = 2;
+
+    public static final int MAX_WRITE_SIZE = 0xfffffff;
+
+    public static final long MAX_SEGMENT_SIZE = 0x0000000fffffffffL;
+
+    final long id;
+
+    @NonNull
+    private FileChannel writeChannel;
+
+    @NonNull
+    private FileChannel readChannel;
+
+    @NonNull
+    String segmentFilePath;
+
+    private boolean isDirty;
+
+    private int refCount = 0;
```

Comment:
I agree. Also `refCount` is a misnomer here, because its not used for GC, but logging/debugging purposes. Also, the Class should be thread-safe according to it's contract irrespective of the locking scheme utilized by `StreamLogFiles`

Reply:
Why do you think the statement is not exactly correct?

This class on it's own is thread safe for readers, without external synchronization.

Reply:
OK. I thought the `concurrency model` you mentioned implies no stale read.